Hw04 auto log

forprogr opened the pull request.

forprogr committed:
- настройка maven для нового модуля

forprogr committed:
- нестабильная версия

forprogr committed:
- нестабильная версия 2

forprogr committed:
- стабильная версия 1

forprogr committed:
Otus. Домашнее задание #4
Реализация задачи:
Автомагическое логирование.
Разработайте такой функционал: метод класса можно пометить самодельной аннотацией @Log, например, так:

class TestLogging {
@Log
public void calculation(int param) {};
}

При вызове этого метода "автомагически" в консоль должны логироваться значения параметров.
Например так.

class Demo {
public void action() {
new TestLogging().calculation(6);
}
}

В консоле дожно быть:
executed method: calculation, param: 6

Обратите внимание: явного вызова логирования быть не должно.

petrelevich commented on hw04-auto-log/src/main/java/ru/forprogr/hw/hw04autolog/descriptions/ClassDescription.java:
```diff
@@ -0,0 +1,38 @@
+public class ClassDescription {
+	private String className;
+
+	private Hashtable<String,MethodDescription> classMethods;
```
Hashtable давно не используется, зачем он вам?

forprogr replied:
Наверное, читал не слишком свежие книги по Java. Переделаю.

petrelevich commented on hw04-auto-log/src/main/java/ru/forprogr/hw/hw04autolog/descriptions/ClassDescription.java:
```diff
@@ -0,0 +1,38 @@
+		return className;
+	}
+
+	public void addClassMethod(MethodDescription p_classMethod){
```
classMethods и className можно в конструкторе заполнить и сделать final?

forprogr replied:
classMethods в конструкторе заполнить не выйдет, так как Asm Core API подразумевает событийную модель нахождения элементов при разборе классов. className - и так заполняется в конструкторе. final - добавлю

petrelevich commented on hw04-auto-log/src/main/java/ru/forprogr/hw/hw04autolog/transformer/CashedClassFileTransformer.java:
```diff
@@ -0,0 +1,93 @@
+
+	public boolean isSkippedClassName(String p_className){
+
+		for( String skippedName : skippedClassNames){
```
используйте contains

forprogr replied:
нет, contains не катит. Тут наверное несколько неверное название метода, суть в том чтоб бы исключить все классы находящиеся в пакете и его подпакетах, а не только иметь возможность исключить класс по имени.

petrelevich commented on hw04-auto-log/src/main/java/ru/forprogr/hw/hw04autolog/transformer/FindMethodLogAnnotations.java:
```diff
@@ -0,0 +1,77 @@
+	@Override
+	public void visitEnd() {
+		if(haveLogAnnotation){
+			classDescription.addClassMethod(methodDescription);
```
Очень долгий и запутанный путь от создания экземпляра classDescription до заполнения его полей.
Так лучше не делать.
В целом очень плохая практика использовать класс как контейнер параметров.

forprogr replied:
"Очень долгий и запутанный путь от создания экземпляра classDescription до заполнения его полей.
Так лучше не делать." - посмотрите как это реализовано в Asm Tree API и скажите где я сделал иначе?
"В целом очень плохая практика использовать класс как контейнер параметров." - а какая хорошая практика? как надо было сделать правильно?

petrelevich replied:
Хорошая практика: константы-параметры на входе - результаты на выходе.

petrelevich commented on hw04-auto-log/src/main/java/ru/forprogr/hw/hw04autolog/transformer/LogClassTransformer.java:
```diff
@@ -0,0 +1,78 @@
+		ClassReader classReader = new ClassReader(p_classByteCodes);
+		ClassWriter classWriter = new ClassWriter(classReader, ClassWriter.COMPUTE_MAXS);
+
+		RenameMethodsWithLog renameMethodsVisitor = new RenameMethodsWithLog(Opcodes.ASM5, classWriter, p_classDescription);
```
Opcodes.ASM5 Не очень хорошо распространять детали ASM по коду ваших классов.
Далее уже совсем не понятно, что такие final int p_api.

forprogr replied:
Да, это косяк, к ночи когда отправлял, уже глаз замылился - поправлю.

petrelevich commented on hw04-auto-log/src/main/java/ru/forprogr/hw/hw04autolog/LogAgent.java:
```diff
@@ -0,0 +1,28 @@
+		System.out.println("\n<i> Start premain(String agentArgs, Instrumentation inst) - ru.forprogr.hw.hw04autolog");
+
+		LogClassTransformer logClassTransformer = new LogClassTransformer();
+		logClassTransformer.addSkippedClassNames("java");
```
Лучше передать сразу коллекцию в конструкторе.

forprogr replied:
Ок. Переделаю.

forprogr committed:
-Правки согласно комментариям в ПуллРеквесте

forprogr committed:
- Все классы заполняются там где они объявлены и там где используются - не изменяются
- Изменен тип аргумента конструктора FindLogAnnotations на ClassWriter
- Рефакторинг имен классов и пакетов
